### PR TITLE
Add config.save.use_pinned_memory_for_d2h to serialization config

### PR DESCRIPTION
**Pull request by mikaylagawarecki**
This was benchmarked with two separate scripts on my A100
(A) Save state_dict of llama3-style model on CUDA to disk with ``torch.save``
(B) Save `ModuleList` of 10 `nn.Linear(10,000, 10,000)` on CUDA to disk with `torch.save`
Timings are an average of 5 runs and benchmark scripts + results are attached

Under both scenarios, we see **~2x speedup in ``torch.save`` time with (``compute_crc32=False`` and ``use_pinned_memory_for_d2h=True``)** compared to the baseline of the current defaults (``compute_crc32=True`` and ``use_pinned_memory_for_d2h=False``

(A)  Save state_dict of llama3-style model on CUDA to disk with ``torch.save`` [[script](https://gist.github.com/mikaylagawarecki/d3a86ea1bb08045d1a839976808d7432)][[results](https://gist.github.com/mikaylagawarecki/f61a4714e5cff703146a1fcb7e0c755c)]

|                                                                                 |  use_pinned_memory_for_d2h=False (Default) |  use_pinned_memory_for_d2h=True |
|-|-|-|
| `compute_crc_32= True`  (Default)| 28.54s | 20.76s |
| `compute_crc_32 = False` | 22.57s |  **14.51s** |

(B) Save `ModuleList` of 10 `nn.Linear(10,000, 10,000)` on CUDA to disk with `torch.save` [[script](https://gist.github.com/mikaylagawarecki/ecbc505436bdd4b5190ef1b3430c12b6)][[results](https://gist.github.com/mikaylagawarecki/4e686bcf030b57de8c3ca74d8f5a88f7)]

|                                                                                 |  use_pinned_memory_for_d2h=False (Default) |  use_pinned_memory_for_d2h=True |
|-|-|-|
| `compute_crc_32= True`  (Default)| 8.38s | 5.53s |
| `compute_crc_32 = False` | 6.94s |  **3.99s** |

Trace of (A) with `use_pinned_memory_for_d2h=True`, `compute_crc32=False`
Screenshot 2024-12-16 at 7 32 33 PM

Baseline trace of (A) with `use_pinned_memory_for_d2h=False`, `compute_crc32=True`
Screenshot 2024-12-16 at 7 38 20 PM

Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #143342
* #143324

### CODE REVIEW

File: torch/serialization.py [hunk 1]
```diff
@@ -1206,7 +1206,13 @@ def persistent_id(self, obj):
             # this means to that to get tensors serialized, you need to implement
             # .cpu() on the underlying Storage
             if storage.device.type != "cpu":
-                storage = storage.cpu()
+                from torch.utils.serialization import config
+
+                if config.save.use_pinned_memory_for_d2h:
+                    storage = storage.to(device="cpu", non_blocking=True)
+                    torch.cuda.current_stream().synchronize()
```
[the remark this answers is not in the document]
Comment (mikaylagawarecki) [reply torch/serialization.py:1213]: I definitely failed to consider the generic accelerator case here :( however, I would kind of disagree that async copy + synchronization is an accurate description here 

(I'm not as familiar with other devices, technically every `tensor.to("cpu", non_blocking=False)` from CUDA is implemented as `cudaMemcypAsync` + `cudaStreamSynchronize`). So imo the real difference here is that the cudaMemcpyAsync is done to pinned memory I think.

File: test/test_serialization.py [hunk 2]
```diff
@@ -4469,6 +4471,43 @@ def __init__(self, t):
             loaded_sd = torch.load(f, weights_only=weights_only)
             self.assertEqual(sd_save, loaded_sd)
 
+    @unittest.skipIf(not torch.cuda.is_available(), "CUDA not available")
```
[the remark this answers is not in the document]
Comment (mikaylagawarecki) [reply test/test_serialization.py]: Changed this test to use the current accelerator, does that work?

File: torch/serialization.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (mikaylagawarecki) [reply torch/serialization.py]: change this code to allocate a pinned memory storage and copy_ into it then synchronize the accelerator

File: docs/source/notes/serialization.rst [hunk 3]
```diff
@@ -503,6 +503,8 @@ Config
 
   * ``compute_crc32``: whether to compute and write the zip file checksum (Default : ``True``).
     See :func:`~torch.serialization.set_crc32_options`.
+  * ``use_pinned_memory_for_d2h``: for storages that are on CUDA when passed to ``torch.save``, whether to
```
Comment (albanD) [inline docs/source/notes/serialization.rst:506]:
```suggestion
  * ``use_pinned_memory_for_d2h``: for storages that are on an accelerator when passed to ``torch.save``, whether to
```

**Review by albanD — COMMENTED** [review]
Thanks for the update, looks good.
I'm surprised you don't need an update in the config file though?

**Review by albanD — APPROVED** [review]

**Comment by mikaylagawarecki** [comment]
Good catch, I [accidentally squashed the config update into the previous PR](https://github.com/pytorch/pytorch/pull/143324/files#diff-030970443ab6ae89bef63cf45fecfaec09d906ac677a736c5129878275681859R20), let me fix that

**Comment by albanD** [comment]
If they land as one, I think that's ok.

**Pull request closed without merging**